tests/main/snap-user-service-socket-activation/task.yaml

```diff
@@ -0,0 +1,40 @@
+summary: Check that sockets for user services are created correctly
+
+systems:
+    # The core snap currently doesn't set /etc/systemd/user as a
+    # writable path, so the test fails
+    - -ubuntu-core-*
+    # Ubuntu 14.04's systemd doesn't have user@.service
+    - -ubuntu-14.04-*
+    # Amazon Linux 2 gives error "Unit user@12345.service not loaded."
+    - -amazon-linux-2-*
+
+kill-timeout: 5m
+
+prepare: |
+    systemctl stop user@"$(id -u test)".service
+
+restore: |
+    systemctl stop user@"$(id -u test)".service
+    rm -rf /run/user/"$(id -u test)"/*
+
+execute: |
+    echo "Create XDG_RUNTIME_DIR for test user"
+    USER_RUNTIME_DIR=/run/user/"$(id -u test)"
+    mkdir -p "$USER_RUNTIME_DIR" || true
+    rm -rf "${USER_RUNTIME_DIR:?}"/*
+    chmod u=rwX,go= "$USER_RUNTIME_DIR"
+    chown test:test "$USER_RUNTIME_DIR"
+
+    echo "When the service snap is installed"
+    #shellcheck source=tests/lib/snaps.sh
+    . "$TESTSLIB"/snaps.sh
+    install_local test-snapd-user-service-sockets
+
+    echo "And the user mode systemd instance is started"
+    systemctl start user@"$(id -u test)".service
+
+    echo "It's sockets are created in the test user's directories"
```

Comment:
Should we do a real end-to-end test here, i.e. have something connect to the socket and verify that the snap actually replies?

Reply:
I've updated the test to use a Python daemon implementation of the service that responds differently to the different sockets.  This could be reused to test socket activation of system services.